cmd/uniq/uniq_test.go

```diff
@@ -0,0 +1,315 @@
+	newStr := new(string)
+	*newStr = s + "\n"
+	return &Line{newStr, ln}
+}
+
+var ttScanLines = []struct {
+	opts     options
+	expected []*Line
+}{
+	{
+		options{printEmptyLines: false},
+		[]*Line{
+			newLine("hello", []int{1, 4}),
+			newLine("world", []int{2}),
+			newLine("世界", []int{6, 7}),
+			newLine("世", []int{8}),
+			newLine("1", []int{9, 14}),
+			newLine("3", []int{10}),
+			newLine("4", []int{11, 13}),
+			newLine("日本語", []int{12}),
+		},
+	},
+	{
+		options{printEmptyLines: true},
+		[]*Line{
+			newLine("hello", []int{1, 4}),
+			newLine("world", []int{2}),
+			newLine("", []int{3, 5}),
+			newLine("世界", []int{6, 7}),
+			newLine("世", []int{8}),
+			newLine("1", []int{9, 14}),
+			newLine("3", []int{10}),
+			newLine("4", []int{11, 13}),
+			newLine("日本語", []int{12}),
+		},
+	},
+}
+
+var ttUniq = []struct {
+	opts     options
+	lines    []*Line
+	expected string
+}{
+	{
+		// uniq
+		options{},
+		ttScanLines[0].expected,
+		`world
+世
+3
+日本語
+`,
+	},
+	{
+		// uniq -num
+		options{printLineNumber: true},
+		ttScanLines[0].expected,
+		`2: world
+8: 世
+10: 3
+12: 日本語
+`,
+	},
+	{
+		// uniq -empty
+		options{printEmptyLines: true},
+		ttScanLines[1].expected,
+		`world
+
+世
+3
+日本語
+`,
+	},
+	{
+		// uniq -empty -num
+		options{
+			printEmptyLines: true,
+			printLineNumber: true,
+		},
+		ttScanLines[1].expected,
+		"2: world\n" +
+			"3,5: \n" +
+			`8: 世
+10: 3
+12: 日本語
+`,
+	},
+	{
+		// uniq -dup
+		options{printDuplicates: true},
+		ttScanLines[0].expected,
+		`hello
+世界
+1
+4
+`,
+	},
+	{
+		// uniq -dup -num
+		options{
+			printDuplicates: true,
+			printLineNumber: true,
+		},
+		ttScanLines[0].expected,
+		`1,4: hello
+6,7: 世界
+9,14: 1
+11,13: 4
+`,
+	},
+	{
+		// uniq -dup -empty
+		options{
+			printDuplicates: true,
+			printEmptyLines: true,
+		},
+		ttScanLines[1].expected,
+		`hello
+
+世界
+1
+4
+`,
+	},
+	{
+		// uniq -dup -empty -num
+		options{
+			printDuplicates: true,
+			printEmptyLines: true,
+			printLineNumber: true,
+		},
+		ttScanLines[1].expected,
+		"1,4: hello\n" +
+			"3,5: \n" +
+			`6,7: 世界
+9,14: 1
+11,13: 4
+`,
+	},
+
+	{
+		// uniq -every
+		options{printEveryOnce: true},
+		ttScanLines[0].expected,
+		`hello
+world
+世界
+世
+1
+3
+4
+日本語
+`,
+	},
+	{
+		// uniq -every -num
+		options{
+			printEveryOnce:  true,
+			printLineNumber: true,
+		},
+		ttScanLines[0].expected,
+		`1,4: hello
+2: world
+6,7: 世界
+8: 世
+9,14: 1
+10: 3
+11,13: 4
+12: 日本語
+`,
+	},
+	{
+		// uniq -every -empty
+		options{
+			printEveryOnce:  true,
+			printEmptyLines: true,
+		},
+		ttScanLines[1].expected,
+		`hello
+world
+
+世界
+世
+1
+3
+4
+日本語
+`,
+	},
+	{
+		// uniq -every -empty -num
+		options{
+			printEveryOnce:  true,
+			printEmptyLines: true,
+			printLineNumber: true,
+		},
+		ttScanLines[1].expected,
+		"1,4: hello\n" +
+			"2: world\n" +
+			"3,5: \n" +
+			`6,7: 世界
+8: 世
+9,14: 1
+10: 3
+11,13: 4
+12: 日本語
+`,
+	},
+}
+
+func cmpLines(a, b []*Line) error {
+	if la, lb := len(a), len(b); la != lb {
```

Comment:
In Go `len` doesn't actually count but only accesses an internal field (`len(s) == s.sz`) and len call is always inlined.
Then, 
```go
if len(a) != len(b) {
         return fmt.Errorf("%T sizes diverge: %v != %v", a, len(a), len(b))
```
isn't slow.
*only a hint, a minor comment.